opm/material/localad/Evaluation.hpp

```diff
@@ -441,11 +443,27 @@ std::ostream& operator<<(std::ostream& os, const Evaluation<Scalar, VarSetTag, n
 // { std::cout << foo(0) << "\n"; }
 //
 // this will print '0' for bar() and '1' for foobar()...
-//
-#error "Due to some C++ peculiarity regarding function template specialization,"
-#error "the 'evaluation.hh' header must be included before Dune's 'densematrix.hh'!"
+#if !(DUNE_VERSION_NEWER(DUNE_COMMON, 2,4))
+
+#include "Math.hpp"
+
+namespace std {
```

Comment:
I'm rather indifferent on the `#warning` vs `#error` issue, so I'll do it as you prefer if you have any preference. my thought was that the error makes it impossible to do things wrong, but if `Evaluation` is used with `Dune::FieldMatrix`, it is a false positive.

you're also right that putting something into `std::` is illegal (strictly speaking), but since this is just a new overload for something which is already in `std::` to hack around an issue in external code, I think that this is not going to be prosecuted too much by the ISO c++ committee ;)